components/autofill/src/db/schema.rs

```diff
@@ -73,76 +75,53 @@ const CREATE_SHARED_SCHEMA_SQL: &str = include_str!("../../sql/create_shared_sch
+    fn define_functions(c: &Connection) -> Result<()> {
+        c.create_scalar_function(
+            "generate_guid",
+            0,
+            FunctionFlags::SQLITE_UTF8,
+            sql_fns::generate_guid,
+        )?;
+        c.create_scalar_function("now", 0, FunctionFlags::SQLITE_UTF8, sql_fns::now)?;
 
-pub fn init(db: &Connection) -> Result<()> {
-    let version = get_current_schema_version(db)?;
-    if version != VERSION {
-        if version < VERSION {
-            upgrade(db, version)?;
-        } else {
-            log::warn!(
-                "Optimistically loaded future schema version {} (we only understand version {})",
-                version,
-                VERSION
-            );
-            // Downgrade the schema version, so that anything added with our
-            // schema is migrated forward when the newer library reads our
-            // database.
-            db.execute_batch(&format!("PRAGMA user_version = {};", VERSION))?;
-        }
-        create(db)?;
+        Ok(())
     }
-    Ok(())
-}
 
-fn create(db: &Connection) -> Result<()> {
-    log::debug!("Creating schema");
-    db.execute_batch(
-        format!(
-            "{}\n{}",
-            CREATE_SHARED_SCHEMA_SQL, CREATE_SHARED_TRIGGERS_SQL
-        )
-        .as_str(),
-    )?;
-    db.execute(
-        &format!("PRAGMA user_version = {version}", version = VERSION),
-        NO_PARAMS,
-    )?;
-    Ok(())
-}
-
-pub fn create_empty_sync_temp_tables(db: &Connection) -> Result<()> {
-    log::debug!("Initializing sync temp tables");
-    db.execute_batch(CREATE_SYNC_TEMP_TABLES_SQL)?;
-    Ok(())
-}
-
-fn upgrade(db: &Connection, from: i64) -> Result<()> {
-    log::debug!("Upgrading schema from {} to {}", from, VERSION);
-    if from == VERSION {
-        return Ok(());
-    }
-    // Places has a cute `migration` helper we can consider using if we get
-    // a few complicated updates, but let's KISS for now.
-    if from == 0 {
+    fn upgrade_from_v0(db: &Connection) -> Result<()> {
         // This is a bit painful - there are (probably 3) databases out there
-        // that have a schema of 0 but actually exist - ie, we can't assume
-        // a schema of zero implies "new database".
+        // that have a schema of 0.
         // These databases have a `cc_number` but we need them to have a
         // `cc_number_enc` and `cc_number_last_4`.
         // This was so very early in the Fenix nightly cycle, and before any
         // real UI existed to create cards, so we don't bother trying to
-        // migrate them, we just drop the table so it's re-created with the
+        // migrate them, we just drop the table and re-create it with the
         // correct schema.
-        db.execute("DROP TABLE IF EXISTS credit_cards_data", NO_PARAMS)?;
-    } else if from == 1 {
+        db.execute_batch(
+            "
+            DROP TABLE IF EXISTS credit_cards_data;
+            CREATE TABLE credit_cards_data (
```

Comment:
The thing I'd like to avoid is re-executing the schema then running more upgrade functions, since the schema that those upgrades work on shouldn't change.   Maybe upgrade functions could return a value that indicates that the database should be deleted and recreated to handle situations like this one.